lxd/device/gpu_sriov.go

```diff
@@ -83,25 +83,49 @@ func (d *gpuSRIOV) Start() (*deviceConfig.RunConfig, error) {
 	saveData := make(map[string]string)
 
 	// Get SRIOV parent, i.e. the actual GPU.
-	parentPCIAddress, err := d.getParentPCIAddress()
+	parentPCIAddresses, err := d.getParentPCIAddresses()
 	if err != nil {
 		return nil, err
 	}
 
-	// Get PCI information about the GPU device.
-	devicePath := filepath.Join("/sys/bus/pci/devices", parentPCIAddress)
+	var parentPCIAddress string
+	var pciParentDev pcidev.Device
+	vfID := -1
 
-	pciParentDev, err := pcidev.ParseUeventFile(filepath.Join(devicePath, "uevent"))
-	if err != nil {
-		return nil, errors.Wrapf(err, "Failed to get PCI device info for GPU %q", parentPCIAddress)
+	// Since there might be multiple GPUs, we iterate through them and get the first free
+	// virtual function.
+	for _, parentPCIAddress = range parentPCIAddresses {
+		// Get PCI information about the GPU device.
+		devicePath := filepath.Join("/sys/bus/pci/devices", parentPCIAddress)
+
+		pciParentDev, err = pcidev.ParseUeventFile(filepath.Join(devicePath, "uevent"))
+		if err != nil {
+			err = errors.Wrapf(err, "Failed to get PCI device info for GPU %q", parentPCIAddress)
+			continue
+		}
+
+		vfID, err = d.findFreeVirtualFunction(pciParentDev)
+		if err != nil {
+			err = errors.Wrap(err, "Failed to find free virtual function")
+			continue
+		}
+
+		if vfID == -1 {
+			continue
+		}
+
+		break
+	}
+
+	if vfID == -1 {
+		return nil, fmt.Errorf("All virtual functions on parent device seem to be in use")
 	}
 
-	vfID, err := d.findFreeVirtualFunction(pciParentDev)
 	if err != nil {
```

Comment:
@monstermunchkin I think this error check block needs to be moved above the vfID == -1 block otherwise it'll never trigger (as if err != nil then vfID should always be -1).

Reply:
and that should be fine now that you're using a single `err` variable, as that last successful call to `findFreeVirtualFunction` should have reset that to nil.

Reply:
Actually looking at `findFreeVirtualFunction` it returns 0 and an error if no VFs are found, so probably don't need that check anyway, but if you do leave it in then, probably worth updating `findFreeVirtualFunction` to return -1 when no VFs found and no nil error.

Reply:
> Actually looking at `findFreeVirtualFunction` it returns 0 and an error if no VFs are found, so probably don't need that check anyway, but if you do leave it in then, probably worth updating `findFreeVirtualFunction` to return -1 when no VFs found and no nil error.

I did that already. If no VFs are found, it'll return `-1, nil`.

Reply:
@monstermunchkin so you did, I'd missed that, sorry.